File: python/sublist/sublist.py
"""
This exercise stub and the test suite contain several enumerated constants.

Enumerated constants can be done with a NAME assigned to an arbitrary,
but unique value. An integer is traditionally used because it’s memory
efficient.
It is a common practice to export both constants and functions that work with
those constants (ex. the constants in the os, subprocess and re modules).

You can learn more here: https://en.wikipedia.org/wiki/Enumerated_type
"""

# Possible sublist categories.
# Change the values as you see fit.
SUBLIST = 0
SUPERLIST = 1
EQUAL = 2
UNEQUAL = 3


def sublist(list_one, list_two):
    if list_one == list_two:
        return EQUAL
    if is_sublist(list_two, list_one):
        return SUPERLIST
    if is_sublist(list_one, list_two):
        return SUBLIST

    return UNEQUAL


def is_sublist(list1, list2):
    length1 = len(list1)
    lenght2 = len(list2)
    return any(list2[i:i + length1] == list1
               for i in range(lenght2 - length1 + 1))

File: c/sublist/sublist.c
#include "sublist.h"
#include <stdbool.h>

static bool is_equal(int *list1, int *list2, size_t length){
    for (size_t index = 0; index < length; ++index){
        if (list1[index] != list2[index])
            return false;
    }
    return true;
}

static bool is_sublist(int *list1, int *list2, size_t length1, size_t length2){
    size_t index1 = 0, index2 = 0;
    while (index2 < length2 && index1 < length1){
        if (list2[index2] == list1[index1])
            ++index1;
        else if (index1) {
            index2 -= index1;
            index1 = 0;
        }
        ++index2;
    }
    return index1 == length1;
}

comparison_result_t check_lists(int *list1, int *list2, size_t length1,
                                size_t length2){
    if (length1 == length2){
        if (is_equal(list1, list2, length1))
            return EQUAL;
        else
            return UNEQUAL;
    }

    if (length1 > length2 && is_sublist(list2, list1, length2, length1))
        return SUPERLIST;

    if (length1 < length2 && is_sublist(list1, list2, length1, length2))
        return SUBLIST;

    return UNEQUAL;
}
